ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+
+    extract_option = fn(opt_name) ->
+      loqui_opts
+        |> Keyword.get(opt_name, [])
+        |> Enum.into(%{}, &{&1.name, &1})
+    end
+
+    registered_codecs = Map.merge(@default_codecs, extract_option.(:codecs))
+    registered_compressors = Map.merge(@default_compressors, extract_option.(:compressors))
+
+    {connect_timeout, tcp_opts} = Keyword.pop(tcp_opts, :connect_timeout, @default_timeout)
+
+    state = %State{
+      host: to_host(host),
+      port: port,
+      tcp_opts: tcp_opts,
+      loqui_path: loqui_path,
+      connect_timeout: connect_timeout,
+      sequence: 1,
+      registered_compressors: registered_compressors,
+      registered_codecs: registered_codecs,
+      buffer: <<>>}
+
+    {:connect, :init, state}
+  end
+
+  def close(conn, timeout \\ 5000),
+    do: Connection.call(conn, :close, timeout)
+
+  def ping(conn, timeout \\ 5000) do
+    Connection.call(conn, :ping, timeout)
+  end
+
+  def request(conn, payload, timeout \\ 5000) do
+    Connection.call(conn, {:request, payload}, timeout)
+  end
+
+  def push(conn, payload) do
+    Connection.cast(conn, {:push, payload})
+  end
+
+  def connect(_info, %{sock: nil, host: host, port: port, tcp_opts: tcp_opts, connect_timeout: connect_timeout}=state) do
+    with {:ok, sock}  <- :gen_tcp.connect(host, port, tcp_opts, connect_timeout),
+         {:ok, state} <- do_upgrade(%{state | sock: sock}),
+         state        <- make_active_once(state),
+         {:ok, state} <- do_loqui_connect(state) do
+
+      {:ok, state}
+    else
+      {:error, _} = error ->
+        Logger.error("Couldn't connect to #{host}:#{port} because #{inspect error}")
+        {:stop, error, state}
+    end
+  end
+
+  def disconnect(info, %State{sock: sock}=state) do
+    :ok = :gen_tcp.close(sock)
+    case info do
+      {:close, from} ->
+        Connection.reply(from, :ok)
+        {:stop, :normal, %{state | sock: nil}}
+
+      {:error, reason}  ->
+        {:stop, reason, %{state | sock: nil}}
+    end
+  end
+
+  # Connection callbacks
+  def handle_call(:ping, caller, %{sock: sock}=state) do
+    {next_seq, state} = next_sequence(state)
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
+  end
+
+  def handle_call(_, _from, %State{sequence: :go_away}=state) do
+    {:reply, {:error, :remote_went_away}, state}
+  end
+
+  def handle_call(:close, from, %State{sock: sock}=s) do
+    go_away_packet = Frames.goaway(0, 0, "Closing")
+    :gen_tcp.send(sock, go_away_packet)
+
+    {:disconnect, {:close, from}, s}
+  end
+
+  def handle_call({:request, payload}, caller, %State{sock: sock, codec: codec}=state) do
+    {next_seq, state} = next_sequence(state)
+    encoded_payload = codec.encode(payload)
+    :gen_tcp.send(sock, Frames.request(0, next_seq, encoded_payload))
+
+    {:noreply, %State{state | waiters: Map.put(state.waiters, next_seq, caller)}}
```

Comment:
We do, when when a response comes in, or a ping is returned. If the server dies, I suppose this could be a memory leak. Even if we monitor, there's no guarantee that the waiters will die; they too could be long lived.